Main.hs

```diff
@@ -2,37 +2,115 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Distribution.Compat.ResponseFile (expandResponse)
+import Text.XHtml (base, abbr, reset)
+
+data Expr = Num Double
+          | Sqrt Expr
+          | Add Expr Expr
+          | Sub Expr Expr
+          | Mul Expr Expr
+          | Div Expr Expr
+          | Pow Expr Expr
 
-data Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num x) = show x
+  show (Sqrt expr) = "sqrt(" ++ show expr ++ ")"
+  show (Add expr1 expr2) = "(" ++ show expr1 ++ ") + (" ++ show expr2 ++ ")"
+  show (Sub expr1 expr2) = "(" ++ show expr1 ++ ") - (" ++ show expr2 ++ ")"
+  show (Mul expr1 expr2) = "(" ++ show expr1 ++ ") * (" ++ show expr2 ++ ")"
+  show (Div expr1 expr2) = "(" ++ show expr1 ++ ") / (" ++ show expr2 ++ ")"
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  a == b = show a == show b
+
 
-data Error
+data Error = ZeroDivision
+           | NegativeNumberSqrt
 
 instance Show Error where
-  show = undefined
+  show :: Error -> String
+  show ZeroDivision = "Division by zero error"
+  show NegativeNumberSqrt = "Taking square root from a negative number error"
+
 
 instance Eq Error where
-  (==) = undefined
+  ZeroDivision == ZeroDivision = True
+  NegativeNumberSqrt == NegativeNumberSqrt = True
+  _ == _ = False
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Num a) = Right a
+eval (Sqrt exp) = let res = eval exp in case res of
+  Right x -> if x >= 0 then Right (sqrt x) else Left NegativeNumberSqrt
+  Left x -> Left x
+eval (Add exp1 exp2) = let
+                  a = eval exp1
+                  b = eval exp2
+                in case (a, b) of
+                  (Right a, Right b) -> Right (a + b)
+                  (Left a, _) -> Left a
+                  (_, Left b) -> Left b
+eval (Sub exp1 exp2) = let
+                  a = eval exp1
+                  b = eval exp2
+                in case (a, b) of
+                  (Right a, Right b) -> Right (a - b)
+                  (Left a, _) -> Left a
+                  (_, Left b) -> Left b
+eval (Mul exp1 exp2) = let
+                  a = eval exp1
+                  b = eval exp2
+                in case (a, b) of
+                  (Right a, Right b) -> Right (a * b)
+                  (Left a, _) -> Left a
+                  (_, Left b) -> Left b
```

Comment:
There is a lot of code repetition which you can get rid of by introducing a helper function.